Add guide for Grafana dashboards

mhajas opened the pull request:
Closes #37129

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mhajas committed:
Add guide for Grafana dashboards

Closes #37129

Signed-off-by: Michal Hajas <mhajas@redhat.com>

ahus1 commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+. Identify the branch from `keycloak-grafana-dashboards` to use from the following table.
++
+
+|===
```
Maybe this should be described in the README.md of the repository as this docs might be eventually for an older release, and then `main` would have continued to evolve for a different release.

mhajas replied:
I was thinking about this as well. My reasoning to put it here is that I assume it won't change very often and would be better to have everything in one place rather than documenting it in the repo. We most probably won't make any breaking changes at least until KC27. Also, I think the way how the mapping works may change in the future, for example, we will have different mapping per dashboard. Do you still think it makes sense to move it?

andymunro replied:
Thanks, I think it's okay to keep it here for now.

ahus1 committed:
Review

Closes #37129

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (dismissed):
Thank you for this pull request! I've done some text editing in https://github.com/mhajas/keycloak/pull/41, mainly adding "." at the end of the enumeration of the steps, and all controls are formatted in bold (AFAIK we only use the typewriter style for file names of text a users puts into a text field). 

See below for some information about branches which IMHO should be moved to the repository as only that repo would have the latest information which branch to use for which version of Keycloak.

mhajas commented:
Thanks @ahus1! I added your changes to this PR and responded to your suggestion.

ahus1 commented:
@andymunro - this is a new guide. Can you please review it when you have the time? Thanks!

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+* {project_name} metrics are enabled. Follow <@links.observability id="configuration-metrics"/> {section} for more details.
+* Grafana instance is running and {project_name} metrics are collected into a Prometheus instance.
+
+== Keycloak Grafana dashboards
```
```suggestion
== {project_name} Grafana dashboards
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+----
+. The dashboards are available in the directory `keycloak-grafana-dashboard/dashboards`.
+
+Below is the list of available dashboards and their purpose.
```
```suggestion
The following section describes the purpose of the available dashboards.
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+
+This dashboard is available in the JSON file: `keycloak-troubleshooting-dashboard.json`.
+
+On the top of the dashboard, there are graphs displaying the service level indicators as defined in <@links.observability id="keycloak-service-level-indicators"/>.
```
```suggestion
On the top of the dashboard, graphs display the service level indicators as defined in <@links.observability id="keycloak-service-level-indicators"/>.
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+This dashboard is available in the JSON file: `keycloak-troubleshooting-dashboard.json`.
+
+On the top of the dashboard, there are graphs displaying the service level indicators as defined in <@links.observability id="keycloak-service-level-indicators"/>.
+It can be also used while troubleshooting {project_name} deployment following the <@links.observability id="metrics-for-troubleshooting"/> {section}, for example, when SLI graphs do not show expected results.
```
```suggestion
This dashboard can be also used while troubleshooting a {project_name} deployment following the <@links.observability id="metrics-for-troubleshooting"/> {section}, for example, when SLI graphs do not show expected results.
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+On the top of the dashboard, there are graphs displaying the service level indicators as defined in <@links.observability id="keycloak-service-level-indicators"/>.
+It can be also used while troubleshooting {project_name} deployment following the <@links.observability id="metrics-for-troubleshooting"/> {section}, for example, when SLI graphs do not show expected results.
+
+A screenshot of this dashboard:
+
+image::observability/keycloak-troubleshooting-grafana-dashboard.png[]
```
```suggestion
.Keycloak troubleshooting dashboard
image::observability/keycloak-troubleshooting-grafana-dashboard.png[]
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+For example, it shows the number of password validations or login flows performed by {project_name}.
+To learn more about these metrics proceed to the following {section} <@links.observability id="metrics-for-troubleshooting-keycloak"/>.
+
+WARNING: It is necessary to enable {project_name} event metrics for this dashboard to work correctly. Proceed to the following {section} <@links.observability id="event-metrics"/> to enable them.
+
+A screenshot of this dashboard:
+
```
```suggestion
NOTE:  {project_name} event metrics must be enabled for this dashboard to work correctly.  To enable them, see the {section} <@links.observability id="event-metrics"/>.

.Keycloak capacity planning dashboard
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+
+. Open the dashboard page from the left Grafana menu.
+. Click *New* and *Import*.
+. Click *Upload dashboard JSON file* and pick the JSON file of the dashboard you want to import.
```
```suggestion
. Click *Upload dashboard JSON file* and select the JSON file of the dashboard you want to import.
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+Follow these steps to export a dashboard to JSON format.
+
+. Open a dashboard you would like to export.
+. Click the share button in the top left corner next to the name of the dashboard.
+. Change tab to *Export*.
+. Enable the option *Export for sharing externally*.
```
```suggestion
. Click *share* in the top left corner next to the dashboard name.
. Click the *Export* tab.
. Enable *Export for sharing externally*.
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+. Click the share button in the top left corner next to the name of the dashboard.
+. Change tab to *Export*.
+. Enable the option *Export for sharing externally*.
+. Then either click on *Save to file* or click on *View JSON* and *Copy to Clipboard* depending on where you want to store the resulting JSON.
```
```suggestion
. Click either *Save to file* or *View JSON* and *Copy to Clipboard* based on where you want to store the resulting JSON.
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+
+Below is the list of available dashboards and their purpose.
+
+=== Keycloak troubleshooting dashboard
```
```suggestion
=== {project_name} troubleshooting dashboard
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+
+== Import a dashboard
+
+To import a dashboard, follow these steps:
+
```
```suggestion
```

andymunro commented on docs/guides/observability/grafana-dashboards.adoc:
```diff
@@ -0,0 +1,91 @@
+
+== Export a dashboard
+
+Exporting a dashboard may be useful, for example, if you want to suggest a change in our dashboard repository.
+Follow these steps to export a dashboard to JSON format.
+
```
```suggestion
Exporting a dashboard to JSON format may be useful. For example, you may want to suggest a change in our dashboard repository.

```

andymunro reviewed (changes requested):
Looks good @mhajas.  I have a few suggestions.

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas committed:
Update docs/guides/observability/grafana-dashboards.adoc

Co-authored-by: andymunro <48995441+andymunro@users.noreply.github.com>
Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas commented:
Thank you for the review @andymunro! I applied all of your comments. Could you please check again? I would like to add one more commit so @ahus1 please wait with merging.

mhajas committed:
Add Grafana share icon to avoid confusion

Signed-off-by: Michal Hajas <mhajas@redhat.com>

mhajas commented:
I added one more commit because:
```
Click *share* in the top left corner next to the dashboard name.
```
wasn't really correct as people need to click share icon, there is no `share` text there. I added an icon into the text and this is how it looks like on the page at the moment:
![image](https://github.com/user-attachments/assets/c8737a37-58c9-4dcb-8d48-8b40c9fd46f1)


I needed to do some magic because `.kc-article` class is adding margin top and bottom to all images and it didn't look very nice to have gave below and above the item in the list.

andymunro reviewed (approved):
Looks good. Thanks, @mhajas

mhajas commented:
Thanks again @andymunro, @ahus1 this should be now ready for your final review and merge.

ahus1 commented:
@andymunro - thank you for the review. It is usually a mixed feeling of embarrassment and relief seeing you fix and improve the texts we write here.
